app/models/workers_scaling/algorithm_runner.rb

```diff
@@ -3,96 +3,69 @@ module WorkersScaling
-    # Otherwise returns nil
-    def self.get(experiment_id)
-      LOGGER.debug "Getting runner from cache, experiment id: #{experiment_id}"
-      return @@cache[experiment_id.to_s] if @@cache
-    end
+    RUNNER_INTERVAL = 15.seconds
+    THREADS_NUMBER = 4
 
     ##
-    # Registers instance of AlgorithmRunner for experiment_id
-    def self.put(experiment_id, runner)
-      @@cache[experiment_id.to_s] = runner
-      LOGGER.debug "Adding runner to cache, experiment id: #{experiment_id}"
-    end
-
-    ##
-    # Unregisters instance of AlgorithmRunner for experiment_id
-    def self.delete(experiment_id)
-      @@cache.delete(experiment_id.to_s) if @@cache
-      LOGGER.debug "Deleting runner from cache, experiment id: #{experiment_id}"
-    end
-
-    ##
-    # Expects:
-    # * instance of Experiment
-    # * instance of Algorithm for given Experiment
-    # * interval between decision loop executions in seconds
-    def initialize(experiment, algorithm, interval)
-      @experiment = experiment
-      @algorithm = algorithm
-      @interval = interval
-      @mutex = Mutex.new
-
-      self.class.put(@experiment.id, self)
-    end
-
-    ##
-    # Initializes Algorithm, then starts its decision loop
-    # Stops when there is no next execution time set
-    def start
+    # Runs periodically and loads some data from all saved Algorithms
+    # If next_execution_time has passed, starts execute_and_schedule for this algorithm
+    def self.start
       Thread.new do
         begin
-          @algorithm.initial_deployment
-          @next_execution_time = Time.now + @interval
-
-          @mutex.synchronize do
-            catch :finished do
-              until @next_execution_time.nil?
+          while true do
+            LOGGER.debug 'Entering Algorithm Runner loop'
+            work_queue = Queue.new
+            Algorithm.get_all_algorithms_times.each do |experiment_id, next_execution_time|
+              if next_execution_time <= Time.now
+                work_queue.push(experiment_id)
+              end
+            end
 
-                while @next_execution_time > Time.now do
-                  @mutex.sleep @next_execution_time - Time.now
-                  throw :finished if @next_execution_time.nil?
+            threads = (1..THREADS_NUMBER).map do
+              Thread.new do
+                until work_queue.empty?
+                  execute_and_schedule(work_queue.pop)
                 end
-
-                execute_and_schedule
               end
             end
-          end
 
-          self.class.delete(@experiment.id)
+            threads.each &:join
+
+            sleep(RUNNER_INTERVAL)
+          end
         rescue => e
-          LOGGER.error "Exception occurred during workers scaling algorithm: #{e.to_s}\n#{e.backtrace.join("\n")}"
+          LOGGER.error "Exception occurred during Algorithm Runner loop: #{e.to_s}\n#{e.backtrace.join("\n")}"
```

Comment:
@mpaciore - taki rekord powinien być oznaczony jako błędny i już nie ruszany (najlepiej usuwany po jakimś czasie po wystąpieniu błędu). Podobnie jak jest w monitoringu SiM.
> w bazie pojawi się rekord, który nie ma experiment_id albo next_execution_time
to też jest błąd, który powinien wyeliminować rekord z algorytmem z dalszego obiegu.

``experiment_status_check`` - odnosi się do pojedycznego algorytmu działającego na konkretnym eksperymencie? Coś jak ``monitor`` dla ``SimulationManager``?

Można by się pokusić o liczbę możliwych wyjątków (np. licznik errorów), ale dla szybszego wykonania teraz zrób oznaczanie rekordu jako error przy pierwszym wyjątku - wyjątki występują rzadko i jeśli już wystąpią to najprawdopodobniej wystąpi ten sam drugi raz.